manta-pay/src/crypto/hash/poseidon/mod.rs

```diff
@@ -0,0 +1,636 @@
+            Fp(lhs.0 * rhs.0)
+        }
+
+        #[inline]
+        fn add_assign(lhs: &mut Self::Field, rhs: &Self::Field, _: &mut ()) {
+            lhs.0 += rhs.0;
+        }
+
+        #[inline]
+        fn add_const_assign(lhs: &mut Self::Field, rhs: &Self::ParameterField, _: &mut ()) {
+            lhs.0 += rhs.0;
+        }
+
+        #[inline]
+        fn apply_sbox(point: &mut Self::Field, _: &mut ()) {
+            point.0 = point.0.pow(&[Self::SBOX_EXPONENT, 0, 0, 0]);
+        }
+    }
+
+    impl<S> super::Specification<Compiler<S>> for S
+    where
+        S: Specification,
+    {
+        type Field = FpVar<S::Field>;
+        type ParameterField = Fp<S::Field>;
+
+        const PARTIAL_ROUNDS: usize = S::PARTIAL_ROUNDS;
+        const FULL_ROUNDS: usize = S::FULL_ROUNDS;
+
+        #[inline]
+        fn domain_tag(arity: usize, compiler: &mut Compiler<S>) -> Self::Field {
+            FpVar::new_witness(compiler.cs.clone(), || {
+                Ok(S::Field::from(((1 << arity) - 1) as u64))
+            })
+            .unwrap()
+        }
+
+        #[inline]
+        fn add(lhs: &Self::Field, rhs: &Self::Field, _: &mut Compiler<S>) -> Self::Field {
+            lhs + rhs
+        }
+
+        #[inline]
+        fn add_const(
+            lhs: &Self::Field,
+            rhs: &Self::ParameterField,
+            _: &mut Compiler<S>,
+        ) -> Self::Field {
+            lhs + FpVar::Constant(rhs.0)
+        }
+
+        #[inline]
+        fn mul(lhs: &Self::Field, rhs: &Self::Field, _: &mut Compiler<S>) -> Self::Field {
+            lhs * rhs
+        }
+
+        #[inline]
+        fn mul_const(
+            lhs: &Self::Field,
+            rhs: &Self::ParameterField,
+            _: &mut Compiler<S>,
+        ) -> Self::Field {
+            lhs * FpVar::Constant(rhs.0)
+        }
+
+        #[inline]
+        fn add_assign(lhs: &mut Self::Field, rhs: &Self::Field, _: &mut Compiler<S>) {
+            *lhs += rhs;
+        }
+
+        #[inline]
+        fn add_const_assign(
+            lhs: &mut Self::Field,
+            rhs: &Self::ParameterField,
+            _: &mut Compiler<S>,
+        ) {
+            *lhs += FpVar::Constant(rhs.0)
+        }
+
+        #[inline]
+        fn apply_sbox(point: &mut Self::Field, _: &mut Compiler<S>) {
+            *point = point
+                .pow_by_constant(&[Self::SBOX_EXPONENT])
+                .expect("Exponentiation is not allowed to fail.");
+        }
+    }
+
+    impl<S, const ARITY: usize> Constant<Compiler<S>> for super::Hasher<S, ARITY, Compiler<S>>
+    where
+        S: Specification,
+    {
+        type Type = super::Hasher<S, ARITY>;
+
+        #[inline]
+        fn new_constant(this: &Self::Type, compiler: &mut Compiler<S>) -> Self {
+            let _ = compiler;
+            Self {
+                additive_round_keys: this.additive_round_keys.clone(),
+                mds_matrix: this.mds_matrix.clone(),
+            }
+        }
+    }
+}
```

Comment:
I don't see any hardcoded test for Poseidon here? How to ensure this implementation is correct?